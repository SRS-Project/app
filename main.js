const electron = require('electron')
const url = require('url')
const path = require('path')
// Module to control application life.
const app = electron.app
// Module to create native browser window.
const BrowserWindow = electron.BrowserWindow

// Keep a global reference of the window object, if you don't, the window will
// be closed automatically when the JavaScript object is garbage collected.
const windows = []
function createWindow() {
  // Create the browser window.
  const width = 1000
  const height = 800
  let w = new BrowserWindow({ width, height })

  // and load the index.html of the app.
  w.loadURL(url.format({
    pathname: path.join(__dirname, 'web', 'index.html'),
    protocol: 'file:',
    slashes: true,
  }))

  windows.push(w)

  // Emitted when the window is closed.
  w.on('closed', () => {
    // Dereference the window object
    windows.splice(windows.indexOf(w), 1)
    w = null
  })
}

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', () => {
  createWindow()
  console.log(app.getPath('appData'))
})

// Quit when all windows are closed.
app.on('window-all-closed', () => {
  // On OS X it is common for applications and their menu bar
  // to stay active until the user quits explicitly with Cmd + Q
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', () => {
  // On OS X it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (windows.length === 0) {
    createWindow()
  }
})
