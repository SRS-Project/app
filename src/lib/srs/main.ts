export interface GradeRecord {
  // Data of the grading itself
  timestamp: Date
  score: Score

  // Data after calculating review
  ease: number
}

export interface Reviewable {
  currentEase: number
  nextReviewData: Date
  gradeHistory: GradeRecord[]
}

// Range should be between 0..1.0 as it's a percentage
type Score = number

export interface SRS {
  grade(item: Reviewable, score: Score): void
}

export class DumbSRS implements SRS {
  grade(item: Reviewable, score: Score) {
    console.log('test')
  }
}
