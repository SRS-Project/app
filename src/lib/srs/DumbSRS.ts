import { SRSLibrary, Reviewable, Score } from 'src/lib/srs/main'

export class DumbSRS implements SRSLibrary {
  private calculateNextReviewDate(item: Reviewable, score: Score): Date {
    const nextDate = new Date()
    nextDate.setDate(nextDate.getDate() + 6 * score + 1)

    return nextDate
  }

  grade(item: Reviewable, score: Score): Reviewable {
    const nextReviewDate = this.calculateNextReviewDate(item, score)
    const reviewedDate = new Date()

    return {
      ...item,
      nextReviewDate,
      gradeHistory: [...item.gradeHistory, { ease: item.currentEase, nextReviewDate, score, reviewedDate }],
    }
  }
}
