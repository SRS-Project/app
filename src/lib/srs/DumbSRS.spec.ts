import { DumbSRS } from 'src/lib/srs/DumbSRS'
import { Reviewable, Score, GradeRecord } from 'src/lib/srs/main'

import * as MockDate from 'mockdate'

describe('DumbSRS', () => {
  const srs = new DumbSRS()
  const item: Reviewable = {
    currentEase: 2.8,
    nextReviewDate: new Date(),
    gradeHistory: [] as GradeRecord[],
  }

  beforeAll(() => {
    // May 19 2018
    MockDate.set(1526713164815)
  })

  afterAll(() => {
    MockDate.reset()
  })

  it('Adds seven days to the next review date if correct', () => {
    const score: Score = 1
    const gradedItem: Reviewable = srs.grade(item, score)
    const currentDate = new Date()

    expect(gradedItem.currentEase).toBe(item.currentEase)
    expect(gradedItem.nextReviewDate.toDateString()).toBe('Sat May 26 2018')

    expect(gradedItem.gradeHistory.length).toBe(1)
    expect(gradedItem.gradeHistory[0].score).toBe(score)
    expect(gradedItem.gradeHistory[0].reviewedDate).toEqual(currentDate)
    expect(gradedItem.gradeHistory[0].ease).toBe(item.currentEase)
    expect(gradedItem.gradeHistory[0].nextReviewDate.toDateString()).toBe('Sat May 26 2018')
  })

  it('Adds 1 day to the next review date if incorrect', () => {
    const gradedItem: Reviewable = srs.grade(item, 0)
    const currentDate = new Date()

    expect(gradedItem.nextReviewDate.toDateString()).toBe('Sun May 20 2018')
  })
})
