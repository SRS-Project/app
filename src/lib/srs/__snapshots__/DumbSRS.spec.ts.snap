// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbSRS Adds 1 day to the next review date if incorrect 1`] = `
Object {
  "currentEase": 2.8,
  "gradeHistory": Array [
    Object {
      "ease": 2.8,
      "nextReviewDate": 2018-05-20T06:59:24.815Z,
      "reviewedDate": 2018-05-19T06:59:24.815Z,
      "score": 0,
    },
  ],
  "nextReviewDate": 2018-05-20T06:59:24.815Z,
}
`;

exports[`DumbSRS Adds seven days to the next review date if correct 1`] = `
Object {
  "currentEase": 2.8,
  "gradeHistory": Array [
    Object {
      "ease": 2.8,
      "nextReviewDate": 2018-05-26T06:59:24.815Z,
      "reviewedDate": 2018-05-19T06:59:24.815Z,
      "score": 1,
    },
  ],
  "nextReviewDate": 2018-05-26T06:59:24.815Z,
}
`;
